File: python/src/parley/budget.py
"""Token budgets (SPEC §8). Replies are fitted to the agent's budget by eliding the largest
arrays/strings; elided parts are parked behind EXPAND handles. Mirrors ts/src/budget.ts."""

from __future__ import annotations

import copy
import os
import time
from typing import Any, Protocol

from ._json import b64url_encode, compact
from .lens import est, lean, lens

MIN_STRING = 200
PLACEHOLDER = "h_XXXXXXXXXXXX"
Path = tuple


class HandleStore(Protocol):
    def put(self, handle: str, value: list | str) -> None: ...
    def get(self, handle: str) -> list | str | None: ...


class MemoryHandleStore:
    """Parked remainders, kept for ``ttl`` seconds (SPEC §4.6 asks for at least 10 minutes)."""

    def __init__(self, ttl: float = 30 * 60):
        self.ttl = ttl
        self._m: dict[str, tuple[list | str, float]] = {}

    def put(self, handle: str, value: list | str) -> None:
        self._m[handle] = (value, time.monotonic() + self.ttl)
        if len(self._m) > 10_000:
            now = time.monotonic()
            self._m = {k: v for k, v in self._m.items() if v[1] > now}

    def get(self, handle: str) -> list | str | None:
        e = self._m.get(handle)
        return e[0] if e and e[1] > time.monotonic() else None


def _get(root: Any, path: Path) -> Any:
    for k in path:
        try:
            root = root[k]
        except (KeyError, IndexError, TypeError):
            return None
    return root


def _set(root: Any, path: Path, value: Any) -> None:
    _get(root, path[:-1])[path[-1]] = value


def _collect(v: Any, path: Path, deep: bool, out: list) -> None:
    if isinstance(v, str):
        if len(v) > MIN_STRING:
            out.append((path, len(v)))
    elif isinstance(v, list):
        if v:
            out.append((path, len(compact(v))))
        if deep:
            for i, x in enumerate(v):
                _collect(x, path + (i,), deep, out)
    elif isinstance(v, dict) and deep:
        for k, x in v.items():
            _collect(x, path + (k,), deep, out)


def _roots(r: dict) -> list[tuple[Path, bool]]:
    """Where elision may happen, and whether it may go inside. Effects, summaries and
    hashes are never elided: the list of proposals may shrink, but a proposal never changes."""
    kind = r.get("kind")
    if kind == "ANSWER":
        return [(("data",), True)]
    if kind == "BRIEF":
        return [(("capabilities",), False)]
    if kind == "PROPOSALS":
        return [(("proposals",), False)] + [(("proposals", i, "data"), True) for i in range(len(r.get("proposals") or []))]
    if kind == "RECEIPT":
        return [(("receipt", "result"), True)]
    return []


def fit(reply: dict, budget: int, store: HandleStore) -> dict:
    """Fit ``reply`` within ``budget`` estimated tokens of Lens. Returns a new reply."""
    if est(lens(reply)) <= budget:
        return reply
    original = copy.deepcopy(reply)
    r = copy.deepcopy(reply)
    cut: dict[Path, int] = {}  # path -> kept length, in elision order
    base_more = list(r.get("more") or [])

    def more_for(real: bool) -> list[dict]:
        out = []
        for path, kept in cut.items():
            if _get(r, path) is None:
                continue  # an ancestor was elided; its remainder carries this
            rest = _get(original, path)[kept:]
            handle = "h_" + b64url_encode(os.urandom(9)) if real else PLACEHOLDER
            if real:
                store.put(handle, rest)
            out.append({
                "handle": handle, "path": ".".join(map(str, path)), "remaining": len(rest),
                "est": est(rest) if isinstance(rest, str) else est(lean(rest)),
            })
        return out

    for _ in range(200):
        more = base_more + more_for(False)
        view = {**r, "more": more} if more else r
        if est(lens(view)) <= budget:
            break
        cands: list = []
        for root, deep in _roots(r):
            v = _get(r, root)
            if v is not None:
                _collect(v, root, deep, cands)
        if not cands:
            break
        path = max(cands, key=lambda c: c[1])[0]  # first of the largest, like the TS reduce
        cur = _get(r, path)
        if isinstance(cur, str):
            already = path in cut
            body = cur[:-1] if already else cur  # strip our "…"
            kept = max(MIN_STRING, len(body) // 2)
            if kept >= len(body):
                break
            cut[path] = min(cut[path], kept) if already else kept
            _set(r, path, body[:kept] + "…")
        else:
            cut[path] = len(cur) // 2 if path not in cut else min(cut[path], len(cur) // 2)
            _set(r, path, cur[: len(cur) // 2])
    more = base_more + more_for(True)
    if more:
        r["more"] = more
    return r

File: ts/src/budget.ts
/**
 * Token budgets (SPEC §8). Replies are fitted to the agent's budget by eliding the
 * largest arrays/strings; elided parts are parked behind EXPAND handles.
 */
import { randomId } from './crypto.js';
import { est, lean, lens } from './lens.js';
import type { More, Reply } from './types.js';

/** `owner`: the holder key allowed to EXPAND it (null for anonymous replies). */
export type Parked = (
  | { kind: 'array'; items: unknown[] }
  | { kind: 'text'; text: string }
) & { owner?: string | null };

export interface HandleStore {
  put(handle: string, value: Parked): void;
  get(handle: string): Parked | undefined;
}

export class MemoryHandleStore implements HandleStore {
  private m = new Map<string, { v: Parked; exp: number }>();
  constructor(private ttlMs = 30 * 60_000) {}
  put(h: string, v: Parked) {
    this.m.set(h, { v, exp: Date.now() + this.ttlMs });

    if (this.m.size > 10_000) {
      for (const [k, e] of this.m) {
        if (e.exp < Date.now()) {
          this.m.delete(k);
        }
      }
    }
  }

  get(h: string) {
    const e = this.m.get(h);

    return e && e.exp > Date.now() ? e.v : undefined;
  }
}

type Path = (string | number)[];
type Container = Record<string | number, unknown>;

const MIN_STRING = 200;

function getAt(root: unknown, path: Path): unknown {
  return path.reduce<unknown>(
    (v, k) => (v == null ? undefined : (v as Container)[k]),
    root,
  );
}

function setAt(root: unknown, path: Path, value: unknown) {
  const parent = getAt(root, path.slice(0, -1)) as Container;

  parent[path[path.length - 1]] = value;
}

interface Candidate {
  path: Path;
  size: number;
}

function collect(v: unknown, path: Path, out: Candidate[]) {
  if (typeof v === 'string') {
    if (v.length > MIN_STRING) {
      out.push({ path, size: v.length });
    }
  } else if (Array.isArray(v)) {
    if (v.length > 0) {
      out.push({ path, size: JSON.stringify(v).length });
    }

    for (const [i, x] of v.entries()) {
      collect(x, [...path, i], out);
    }
  } else if (v && typeof v === 'object') {
    for (const [k, x] of Object.entries(v)) {
      collect(x, [...path, k], out);
    }
  }
}

/**
 * Where elision is allowed per reply kind. `deep` roots may be elided anywhere inside;
 * `list` roots only by dropping whole trailing items, so effects, summaries and hashes of
 * what remains are never altered. Deep roots are elided before lists.
 */
function roots(r: Reply): { deep: Path[]; list: Path[] } {
  switch (r.kind) {
    case 'ANSWER':
      return { deep: [['data']], list: [] };
    case 'BRIEF':
      return { deep: [], list: [['capabilities']] };
    case 'PROPOSALS':
      return {
        deep: r.proposals.map((_, i) => ['proposals', i, 'data']),
        list: [['proposals']],
      };
    case 'RECEIPT':
      return { deep: [['receipt', 'result']], list: [] };
    default:
      return { deep: [], list: [] };
  }
}

/** Elidable strings/arrays in `r`: deep roots first; whole list roots only once none remain. */
function candidates(r: Reply): Candidate[] {
  const { deep, list } = roots(r);
  const cands: Candidate[] = [];

  for (const root of deep) {
    const v = getAt(r, root);

    if (v !== undefined) {
      collect(v, root, cands);
    }
  }

  if (cands.length) {
    return cands;
  }

  for (const root of list) {
    const v = getAt(r, root);

    if (Array.isArray(v) && v.length) {
      cands.push({ path: root, size: JSON.stringify(v).length });
    }
  }

  return cands;
}

/** A value shortened to its first `kept` chars/items, parked behind `handle`. */
interface Cut {
  path: Path;
  kept: number;
  handle: string;
}

interface FitState {
  /** The reply as given, holding the full values that cuts elide. */
  original: Reply;
  /** The reply being shrunk. */
  r: Reply;
  /** Cuts by dotted path. */
  cuts: Map<string, Cut>;
}

/**
 * Halve the string or array at `path`, recording the cut (repeat cuts keep their handle).
 * Returns false when it cannot shrink further.
 */
function halve(s: FitState, path: Path): boolean {
  const cur = getAt(s.r, path);
  const key = path.join('.');
  const prev = s.cuts.get(key);

  if (typeof cur === 'string') {
    const body = prev ? cur.slice(0, -1) : cur; // strip our "…"
    const kept = Math.max(MIN_STRING, Math.floor(body.length / 2));

    if (kept >= body.length) {
      return false;
    }

    s.cuts.set(key, {
      path,
      kept: prev ? Math.min(prev.kept, kept) : kept,
      handle: prev?.handle ?? randomId('h'),
    });
    setAt(s.r, path, `${body.slice(0, kept)}…`);

    return true;
  }

  if (!Array.isArray(cur)) {
    return false;
  }

  const kept = Math.floor(cur.length / 2);

  s.cuts.set(key, { path, kept, handle: prev?.handle ?? randomId('h') });
  setAt(s.r, path, cur.slice(0, kept));

  return true;
}

/** What a cut elided: the rest of the original string or array after `kept`. */
function remainder(full: unknown, kept: number): Parked {
  if (typeof full === 'string') {
    return { kind: 'text', text: full.slice(kept) };
  }

  if (Array.isArray(full)) {
    return { kind: 'array', items: full.slice(kept) };
  }

  throw new TypeError('only strings and arrays are cut');
}

/** One `More` entry per cut still visible in `s.r`; `park` receives each elided remainder. */
function moreEntries(
  s: FitState,
  park?: (handle: string, rest: Parked) => void,
): More[] {
  const out: More[] = [];

  for (const { path, kept, handle } of s.cuts.values()) {
    if (getAt(s.r, path) === undefined) {
      continue; // an ancestor was elided; its remainder carries this
    }

    const rest = remainder(getAt(s.original, path), kept);

    park?.(handle, rest);
    out.push({
      handle,
      path: path.join('.'),
      remaining: rest.kind === 'text' ? rest.text.length : rest.items.length,
      est: rest.kind === 'text' ? est(rest.text) : est(lean(rest.items)),
    });
  }

  return out;
}

const moreOf = (r: Reply): More[] => ('more' in r ? r.more : undefined) ?? [];

/** Fit `reply` within `budget` estimated tokens of Lens. Returns a new reply (input untouched). */
export function fit<R extends Reply>(
  reply: R,
  budget: number,
  store: HandleStore,
  owner: string | null = null,
): R {
  const r = structuredClone(reply);
  const s: FitState = { original: structuredClone(reply), r, cuts: new Map() };
  const baseMore = moreOf(r);

  for (let guard = 0; guard < 200; guard++) {
    const more = [...baseMore, ...moreEntries(s)];
    const text = lens(
      Object.assign({}, r, { more: more.length ? more : undefined }),
    );

    if (est(text) <= budget) {
      break;
    }

    const cands = candidates(r);

    if (!cands.length) {
      break;
    }

    const pick = cands.reduce((a, b) => (b.size > a.size ? b : a));

    if (!halve(s, pick.path)) {
      break;
    }
  }

  const more = [
    ...baseMore,
    ...moreEntries(s, (handle, rest) => store.put(handle, { ...rest, owner })),
  ];

  if (more.length) {
    Object.assign(r, { more });
  }

  return r;
}
